rootfsCloner/utils.go

```diff
@@ -0,0 +1,391 @@
+//
+// Copyright 2022 Nestybox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package rootfsCloner
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	mapset "github.com/deckarep/golang-set"
+	"github.com/opencontainers/runc/libcontainer/mount"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+
+	sh "github.com/nestybox/sysbox-libs/idShiftUtils"
+)
+
+func createNewOvfsDir(info ovfsMntInfo) error {
+	subdirs := []string{info.mergedDir, info.diffDir, info.workDir}
+
+	for _, subdir := range subdirs {
+		if err := os.MkdirAll(subdir, 0755); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func filePerm(path string) (os.FileMode, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return os.FileMode(0), err
+	}
+	return fi.Mode().Perm(), nil
+}
+
+// Creates the cloned rootfs overlayfs mounts and bind-mounts over the orig rootfs.
+func mountClone(ci *cloneInfo) error {
```

Comment:
Neat. Thinking about how to re-use this pattern for other/new features (extra security) ...